openshift.py

```diff
@@ -0,0 +1,750 @@
+            pprint.pprint(role_binding)
+            if (
+                "userNames" in role_binding.keys()
+                and role_binding["userNames"] is not None
+                and user in role_binding["userNames"]
+            ):
+                return True
+        return False
+
+    def get_all_moc_rolebindings(self, user, project_name):
+        rolebindings = []
+        for role in ["admin", "member", "view"]:
+            if self.user_rolebinding_exists(user, project_name, role):
+                role_bindings.push(role)
+        if role_bindings.size() > 0:
+            return Response(
+                response=json.dumps(
+                    {"msg": "role found", "rolebindings": rolebindings}
+                ),
+                status=200,
+                mimetype="application/json",
+            )
+        return Response(
+            response=json.dumps({"msg": "roles not found"}),
+            status=404,
+            mimetype="application/json",
+        )
+
+    def update_user_role_project(self, project_name, user, role, op):
+        # The REST API 'create rolebindings' doesn't work the way that 'oc create rolebindings'
+        # with the REST API,
+        #     GET can be used to get the specific role from the project (or all roles)
+        #     POST the role is bound to the project (and users)
+        #     PUT is used to modify the rolebindings (add users or groups to the role)
+        #     DELETE is used to remove the role (and all users) from the project
+        #
+        # Don't do anything incorrectly as the error messages are generic enough to be meaningless
+        #
+        # First check to see if there is a rolebinding on the project
+        if op not in ["add", "del"]:
+            return Response(
+                response=json.dumps({"msg": "op is not in ('add' or 'del')"}),
+                status=400,
+                mimetype="application/json",
+            )
+        # add_openshift_role(token,self.get_url(),project_name, role)
+
+        openshift_role = None
+        if role == "admin":
+            openshift_role = "admin"
+        elif role == "member":
+            openshift_role = "edit"
+        elif role == "reader":
+            openshift_role = "view"
+        else:
+            return Response(
+                response=json.dumps(
+                    {
+                        "msg": "Error: Invalid role,  "
+                        + role
+                        + " is not one of 'admin', 'member' or 'reader'"
+                    }
+                ),
+                status=400,
+                mimetype="application/json",
+            )
+
+        r = self.get_rolebindings(project_name, openshift_role)
+        # print("A: result: "+r.text)
+        if not (r.status_code == 200 or r.status_code == 201):
+            # try to create the roles for binding
+            r = self.create_rolebindings(project_name, user, openshift_role)
```

Comment:
If I follow the logic here correctly, then when someone calls `update_user_role_project` with `op == "del"` and the corresponding rolebinding *does not* already exist, it will get created. I think this is probably a bug.

Reply:
This is more interesting than what you suspect, though I will add a op != "del" as a defensive measure.

The important line to consider is the one before what you included:

''''
result = self.get_rolebindings(project_name, openshift_role)
''''

Upon inspection of the parameters, there is no user name in the call.  Basically, because of cleaver way that OpenShift works for this specific call, it returns all of the rolls for all of the users on the given project.  So if there there are any role bindings for the project, this call will succeed.

However, this call will fail when it is a new project.  New projects have no initial role bindings.  When a project is created, we are always creating a project admin for the project (doesn't make any sense otherwise).  We never should create a project and then not bother to add any users to it.

See, it is truly cleaver, and makes things more interesting when you add and delete specific users from specific roles.  It is a matter of getting the list of users and roles from the project, iterating through them, removing them from the list and updating the list.  But this can only be done if the list is first created.